untyped-lambda-calculus/article.md

```diff
@@ -0,0 +1,552 @@
+# A Brief Look at Untyped Lambda Calculus
+
+When talking about Haskell, the term "lambda calculus" often crops up.
+It's a theoretical framework used to define the meaning of computation in many functional languages, such as Haskell, Agda, Idris, etc.
+Understanding lambda calculus can be very helpful when talking about programs written in these languages.
+
+Perhaps surprisingly, understanding lambda calculus can also be helpful in understanding the C++ template metalanguage since it's a functional language in its own right (although it's not, strictly speaking, based on lambda calculus).
+It's also used extensively in programming language theory research and functional language design.
+
+There are two broad kinds of lambda calculus: _untyped_ and _typed_.
+The untyped lambda calculus, discussed in this article, expresses an unrestricted computation.
+While it doesn't find much use outside of computability theory, it's necessary to understand it to meaningfully discuss the typed varieties.
+
+## A bit of history
+
+> 1936 - Alan Turing invents every programming language that will ever be but is shanghaied by British Intelligence to be 007 before he can patent them.
+>
+> 1936 - Alonzo Church also invents every language that will ever be but does it better.
+> His lambda calculus is ignored because it is insufficiently C-like.
+> This criticism occurs in spite of the fact that C has not yet been invented.
+>
+> -- James Iry, [A Brief, Incomplete, and Mostly Wrong History of Programming Languages][brief-incomplete-history]
+
+[brief-incomplete-history]: http://james-iry.blogspot.com/2009/05/brief-incomplete-and-mostly-wrong.html
+
+Lambda calculus, initially envisioned as a formal logic system, was developed by [Alonzo Church](https://en.wikipedia.org/wiki/Alonzo_Church) around the 1930s to explore the foundations of mathematics.
+
+The initial formulation had a logical inconsistency known as the [Kleene–Rosser paradox].
```

Comment:
Reference

Reply:
Eh. So here's the deal: I've added the reference to https://plato.stanford.edu/entries/paradoxes-contemporary-logic/#IncoCertFormLogi, but it was removed in 5690f94 by @NaeosPsy, I'm guessing because it's a bit too opaque for a casual reader.

I don't really care either way, it's not a journal publication where every statement has to be either referenced or proved, but I'd like to avoid editing wars :sweat_smile:

Reply:
You still have a link to Kleene-Rosser paradox, right? I just didn't believe anyone that read this would care enough for two links. You can perhaps link the plato in the Kleene-Rosser paradox hyperlink.

Reply:
Okay, let's try this: I've added a journal-style reference in 87137c74d9b2ea5075b6f7d4169037f2b6df4472.